lib/sources.nix

```diff
@@ -229,6 +229,42 @@ let
 
   canCleanSource = src: src ? _isLibCleanSourceWith || !(pathHasContext (toString src));
 
+  /*
+    Get all files (recursively) that pass a filter function. Unlike other source
+    filtering functions, this will not include empty directories in the final
+    derivation.
+
+    Type: sourceLike -> (String -> Bool) -> Source
+
+    Example:
+      filterFiles ./. (lib.hasSuffix ".html")
+  */
+  filterFiles = src: filter:
+    let
+      includedPaths =
+        builtins.filter
+          filter
+          (lib.filesystem.listFilesRecursive src);
+
+      name =
+        # Fix the problem of store paths names causing a rebuild every time they change.
```

Comment:
This problem is not unique to this function, so it should be moved somewhere else.

Reply:
Am I good to pull this out into its own function and submit a PR with both, or should it be pulled out and put into a separate PR?

Reply:
Actually this is no good, because it reintroduces the parent path impurity.
Unfortunately it's up to us to make sure that the evaluation of a directory does not depend on the location of that directory. That location includes the basename: someone may use a different checkout, or a flake source may resolve to a different location depending on the *entire* project through the store path hash, thereby largely defeating the purpose of source filtering.
`name` should be `"source"` unless the caller specifies something else. In this case you can let the caller set the name afterwards with `cleanSourceWith`. @infinisil is working on https://github.com/NixOS/nixpkgs/pull/112083, which will add a more convenient function for setting the name.

Reply:
could you elaborate on what exactly "largely defeats the purpose of source filtering"?

Reply:
Apologies; I might have responded too quickly here. Nonetheless I am not sure if this new solution should be applied to this new function, and whether this new function fits the larger design of the source combinator library. I'll defer to @infinisil who's currently more up to speed about all of that, including lazy trees.

Reply:
@ursi I'm not entirely sure I understand this name handling here. What exactly does it do and what use cases are affected by this?

Reply:
It addresses the problem caused by using `builtins.filterSource` on a generated source. Such a source will have a name of the form `/nix/store/c59y63zj2pgjx87184kv1kyxqqs9b5sl-source`, which will be updated any time that source changes (A particularly common example of this is the source `./.` inside a flake). Because `filterSource` uses the base name of the file for the name of the new derivation, this means that whenever the generated source changes, the source created from `filterSource` changes, even if the files are exactly the same - it has a different name.

This function checks to see if you're filtering a source from a derivation, and if you are, it just uses the generic name "source" instead of getting the name from the base name like `filterSource` would.

Reply:
I have gone ahead and just used the name "source" for everything.